test/controllers/users_controller_test.rb

```diff
@@ -0,0 +1,100 @@
+require "test_helper"
+
+describe UsersController do
+  describe "index" do
+    it "can render" do
+      get users_path
+
+      must_respond_with :ok
+    end
+
+    it "renders even if there are zero users" do
+      User.destroy_all
+
+      get users_path
+
+      must_respond_with :ok
+    end
+  end
+
+  describe "show" do
+    it "returns a 404 status code if the user doesn't exist" do
+      user_id = "FAKE ID"
+
+      get user_path(user_id)
+
+      must_respond_with :not_found
+    end
+
+    it "works for a User instance that exists" do
+      user = users(:one)
+      
+      get user_path(user)
+
+      must_respond_with :ok
+    end
+  end
+
+  describe "login form" do
+    it "can render login_form" do
+      get login_path
+      must_respond_with :success
+    end
+  end
+
+  describe "login" do
+    it "can login user" do
+      expect(perform_login).must_equal users(:one)
+    end
+
+    it "can find an existing user and flash success message" do
+      user_input = {
+        "user": {
+          username: users(:one).username,
```

Comment:
I like these two test cases (new and existing user). Is there a way for login to fail (e.g. the user logs in as "")?